File: interview/amazon/11.container-with-most-water.cpp
/*
 * @lc app=leetcode id=11 lang=cpp
 *
 * [11] Container With Most Water
 */

#include "bits/stdc++.h"

using namespace std;

// @lc code=start
class Solution {
public:
  int maxArea(vector<int> &height) {
    int res = 0;
    int n = height.size();
    int l = 0, r = n - 1;
    while (l < r) {
      res = max(res, min(height[l], height[r]) * (r - l));
      // Accoarding to Cannikin law, 
      if (height[l] < height[r]) {
        ++l;
      } else {
        --r;
      }
    }

    return res;
  }
};

// @lc code=end

// brute force
// Time Limit Exceeded
// 53/60 cases passed (N/A)
class Solution {
public:
  int maxArea(vector<int> &height) {
    int res = 0;
    int n = height.size();
    for (int i = 0; i < n - 1; ++i) {
      for (int j = i + 1; j < n; ++j) {
        res = max(res, (j - i) * min(height[i], height[j]));
      }
    }
    return res;
  }
};

File: 11.container-with-most-water.py
#
# @lc app=leetcode id=11 lang=python3
#
# [11] Container With Most Water
#

# @lc code=start
from typing import List


# O(n)
class Solution:
    def maxArea(self, height: List[int]) -> int:
        res, l, r = 0, 0, len(height)-1
        while l < r:
            res = max(res, min(height[l], height[r])*(r-l))

            if height[l] < height[r]:
                l += 1
            else:
                r -= 1
        return res


# @lc code=end


# O(n**2)
class Solution2:
    def maxArea(self, height: List[int]) -> int:
        n = len(height)
        res = 0
        for i in range(0, n-1):
            for j in range(1, n):
                res = max(
                    res, (j-i)*(min(height[i], height[j]))
                )

        return res
